### PR TITLE
High throughput transport implementation

### LINKED ISSUE #1097: Java Server Transport
The goal is to have a transport with high transaction throughput.

### Constraints and requirements:

* Not required by Android
* Doesn't require offline caching
* Flushed data doesn't need to be sent in order
* Measure from baseline each change as we experiment
* Metrics on dropped items. i.e: Atomic increment for dropped items so we have visibility

### Some ideas:

* Parallelize envelope serialization
* Maintain N concurrent open connections
* Producer/Consumer `Envelope` to `Requests` 
  * `Requests` are gzip compressed blobs ready for the network I/O
* To test it out: [Benchmarks](https://www.baeldung.com/java-microbenchmark-harness) with a fake HTTP simulating a 1ms latency.

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

Provides high throughput transport implementation meant to be used by backend applications using the performance feature.

Extracts rate limiting functionality to separate class, making it possible to use from other implementation of `ITransport`.

## :bulb: Motivation and Context

Fixes #1097.

- extracted rate-limiter that can be re-used in multiple transport implementations
- lifting `TransportFactory` as a top level interface that can be set on `SentryOptions` - to provide an easy way to set custom transports
- Apache Http Client 5 based high throughput transport implementation that under the hood is based on non blocking IO.
- Spring Boot auto-configuration for the new transport

The last two are in the PR pending review (https://github.com/getsentry/sentry-java/pull/1136).

## :green_heart: How did you test it?

Unit & integration tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [ ] No breaking changes

## :crystal_ball: Next steps

- Merge https://github.com/getsentry/sentry-java/pull/1136
- Do a final review for https://github.com/getsentry/sentry-java/pull/1114

### CODE REVIEW

**Review by bruno-garcia — APPROVED** [review]
nice!

**Review by marandaneto — APPROVED** [review]
LGTM

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java [hunk 1]
```diff
@@ -11,57 +11,79 @@
...
+    this.rateLimiter = Objects.requireNonNull(rateLimiter, "rateLimiter is required");
+    this.transportGate = Objects.requireNonNull(transportGate, "transportGate is required");
+    this.connection = Objects.requireNonNull(httpConnection, "httpConnection is required");
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void send(SentryEnvelope envelope, Object hint) throws IOException {
+    // For now no caching on envelopes
+    IEnvelopeCache currentEnvelopeCache = envelopeCache;
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:70]: Does this break Android?

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java [hunk 2]
```diff
@@ -11,57 +11,79 @@
...
+    }
+
+    final SentryEnvelope filteredEnvelope = rateLimiter.filter(envelope, hint);
+
+    if (filteredEnvelope == null) {
+      if (cached) {
+        envelopeCache.discard(envelope);
+      }
+    } else {
+      executor.submit(new EnvelopeSender(filteredEnvelope, hint, currentEnvelopeCache));
```
Comment (bruno-garcia) [inline sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:85]: What happens if we capture in a tight loop, executor drops after maxItems?

**Review by bruno-garcia — APPROVED** [review]

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java (hunk 1 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:70]: Note that this transport is the same old one we had before.

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java (hunk 2 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:85]: Yes, in such case once we reach max queue items are dropped.

File: sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  @Override
+  public @NotNull ITransport create(final @NotNull SentryOptions options) {
+    Objects.requireNonNull(options, "options is required");
+    final Dsn parsedDsn = new Dsn(options.getDsn());
+    final IConnectionConfigurator credentials =
+        new CredentialsSettingConfigurator(parsedDsn, options.getSentryClientName());
+
+    URL sentryUrl;
+    try {
+      sentryUrl = parsedDsn.getSentryUri().toURL();
```
[the remark this answers is not in the document]
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/AsyncHttpTransportFactory.java:24]: It is done with `init`. If DSN is wrong SDK init will fail.

This piece will be refactored anyways in subsequent PR as every transport needs an envelope url, so there is no point for each factory to resolve it by itself.

File: sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java (hunk 1 above)
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/transport/AsyncHttpTransport.java:70]: yeah, the comment is outdated btw

**Comment by bruno-garcia** [comment]
any blockers for merging this?

**Comment by maciejwalkowiak** [comment]
It's a PR for branch that takes all the NIO related changes in. Once https://github.com/getsentry/sentry-java/pull/1136 is approved, this whole feature is complete and we can merge this PR too.

**Comment by maciejwalkowiak** [comment]
https://github.com/getsentry/sentry-java/pull/1136 is merged. Please do a final review of this PR.

**Review by bruno-garcia — APPROVED** [review]

**Pull request merged**